### PR TITLE
BUG: Jaccard distance for all-zero arrays would return np.nan

### LINKED ISSUE #7317: Jaccard distance both zero vectors behaviour
As far as I know, Jaccard distance between two zero boolean vectors should be equal to zero. 

However now implementation returns ```nan``` as a result.

Example code:
```python
import numpy as np
from scipy.spatial.distance import jaccard

a = np.zeros(5)
print jaccard(a, a)
```

This gives result and the warning:
```
RuntimeWarning: invalid value encountered in double_scalars  / 
np.double(np.bitwise_or(u != 0, v != 0).sum()))
nan
```

Maybe returning a zero would be a more acceptable option?
Python 2.7.12
Scipy version 0.18.1

### PR DESCRIPTION
**Pull request by ElDeveloper**
The Jaccard distance returns `np.nan` when two all-zero vectors are
compared. This patch changes this to return 0 instead, and avoid a
division by zero.

Fixes #7317

### CODE REVIEW

**Comment by jakevdp** [comment]
I don't think this behavior is correct: according to the [definition of jaccard](https://docs.scipy.org/doc/scipy-0.19.0/reference/generated/generated/scipy.spatial.distance.jaccard.html), the similarity score between two all-zero vectors is exactly 0/0, the best representation of which is NaN.

**Comment by jakevdp** [comment]
On the other hand, according to [wikipedia](https://en.wikipedia.org/wiki/Jaccard_index), the Jaccard index between two empty sets is equal to one, not to zero or NaN.

**Comment by ElDeveloper** [comment]
> I think the current behavior is correct: according to the definition of jaccard, the similarity score between two all-zero vectors is exactly 0/0, the best representation of which is NaN.

I partially agree with that, my only intuition against it is that the distance between two identical vectors should be zero.

> On the other hand, according to wikipedia, the Jaccard index between two empty sets is equal to one, not to zero or NaN.

This function refers to the distance, not the index though.

----------

In any case, I would love to clear this out as I was a bit surprised to see `np.nan` in some of my distance matrices last week.

` ¯\_(ツ)_/¯`

**Comment by ElDeveloper** [comment]
> Edit: looking closer the Jaccard index is 1 minus the Jaccard distance, so wikipedia agrees with your change. I'm still not totally convinced that this should be special-cased, though. For example, does this modified Jaccard distance still satisfy the triangle inequality?

Good point, I am not sure. Thinking more about this, the

**Comment by ElDeveloper** [comment]
> Edit: looking closer the Jaccard index is 1 minus the Jaccard distance, so wikipedia agrees with your change. I'm still not totally convinced that this should be special-cased, though. For example, does this modified Jaccard distance still satisfy the triangle inequality?

Good point! My original understanding of this metric was given by the [Encyclopedia of Distances](http://www.uco.es/users/ma1fegan/Comunes/asignaturas/vision/Encyclopedia-of-distances-2009.pdf), specifically in page 299:

screen shot 2017-05-02 at 10 49 40 pm

So I think a more general approach to this would be to have all elements (not just non-zero elements) being considered. This would mean that the denominator should always be the length of the input vectors and not the sum of all non-zero elements. However this conflicts with the equation definition in the docstring, and with Wikipedia.

**Comment by ElDeveloper** [comment]
> This would mean that the denominator should always be the length of the input vectors and not the sum of all non-zero elements.

🤔 I'm rather hesitant about this though.

**Comment by ElDeveloper** [comment]
FWIW, Mathematica's [`JaccardDissimilarity`](https://reference.wolfram.com/language/ref/JaccardDissimilarity.html) returns zero for all-zero vectors:

screen shot 2017-05-02 at 11 04 18 pm

**Comment by jakevdp** [comment]
I'd love to see a solid reference that advocates for special-casing the jaccard index/similarity of null sets. I don't consider Wikipedia or Mathematica solid references :smile:

**Comment by ElDeveloper** [comment]
Absolutely! I will look for something. 👍

**Comment by ElDeveloper** [comment]
@jakevdp, I was able to find two sources** that advocate for the special casing. In [Section 1.2](http://matwbn.icm.edu.pl/ksiazki/cm/cm6/cm6141.pdf) they describe the metric with a special case where the denominator is zero. [Here they note](https://arxiv.org/pdf/1612.02696.pdf) that the Jaccard-index is one when the size of the union of the sets is zero, then they define the metric as 1-Jaccard-index. I did not read thoroughly the papers as I think this confirms our speculations and the papers touch on other topics.

Let me know what you think.

----------------

** one of the sources lead me to the other.

**Comment by rgommers** [comment]
@jakevdp any more thoughts on this?

**Comment by tylerjereddy** [comment]
I'm +1 for returning zero when both inputs are empty given the clear definition in the [arxiv paper](https://arxiv.org/pdf/1612.02696.pdf) mentioned above, which has now been cited ~20 times, including by 1 or 2 journals / proceedings I recognize.

The lead author on that paper is a Mathematics Professor in Germany -- I'll ping their github account @synoptiker here. Sven's twitter account is far more active though so I'll ping there too.

Still that paper + Wikipedia article are pretty clear to me on this matter -- 0 / 0 is special cased to 0 because you formally define the index as `1` in the all-zero case and subtract that from `1` in the formal definition of distance.

This could probably do with a doc update / citation of the manuscript if others agree.

**Comment by ElDeveloper** [comment]
Thanks @tylerjereddy, I have included the references in the documentation.

**Comment by synoptiker** [comment]
Though, numerically, NaN is certainly an appropriate representation for zero divided by zero if you don't have any further information, it is not necessary to be so restrictive in the case of the Jaccard distance. 

Indeed, since any set A should have distance zero to itself (equally, any set A is most similar to itself), setting the Jaccard distance J_\delta(\emptyset,\emptyset)=0 (or, the Jaccard similarity index J(\emptyset,\emptyset)=1, resp.) is plausible. Mathematically, everything turns out to work fine (in particular, it satisfy the triangle inequality).

File: scipy/spatial/distance.py [hunk 1]
```diff
@@ -829,6 +829,18 @@ def jaccard(u, v, w=None):
...
+    Notes
+    -----
+    When both `u` and `v` lead to a `0/0` divsion i.e. there is no overlap
+    between the items in the vectors the returned distance is 0. See the
+    Wikipedia page on the Jaccard index [1]_, and this paper [2]_.
+
+    References
+    ----------
+    .. [1] https://en.wikipedia.org/wiki/Jaccard_index
+    .. [2] S. Kosub, "A note on the triangel inequalty for the Jaccard
```
Comment (tylerjereddy) [inline scipy/spatial/distance.py]:
```suggestion
    .. [2] S. Kosub, "A note on the triangle inequality for the Jaccard
```

File: scipy/spatial/distance.py [hunk 2]
```diff
@@ -829,6 +829,18 @@ def jaccard(u, v, w=None):
...
+       dstance", 2016, Available online: https://arxiv.org/pdf/1612.02696.pdf
```
Comment (tylerjereddy) [inline scipy/spatial/distance.py]:
```suggestion
       distance", 2016, Available online: https://arxiv.org/pdf/1612.02696.pdf
```

File: scipy/spatial/distance.py [hunk 3]
```diff
@@ -844,14 +856,16 @@ def jaccard(u, v, w=None):
...
     unequal_nonzero = np.bitwise_and((u != v), nonzero)
     if w is not None:
         w = _validate_weights(w)
         nonzero = w * nonzero
         unequal_nonzero = w * unequal_nonzero
-    dist = np.double(unequal_nonzero.sum()) / np.double(nonzero.sum())
-    return dist
+    a = np.double(unequal_nonzero.sum())
+    b = np.double(nonzero.sum())
+    return (a / b) if b != 0 else 0
```
Comment (tylerjereddy) [inline scipy/spatial/distance.py]: maybe not the most Pythonic way to write it, but fair enough I suppose

**Review by tylerjereddy — APPROVED** [review]
Ok, the author of the cited paper has weighed in in favor of this change as well (thanks @synoptiker), so let's get this merged very soon.

@ElDeveloper can you address my typo fixes for the doc change? Then I think this should be good to go.

**Comment by ElDeveloper** [comment]
Thanks for the reviews, and comments @tylerjereddy and @synoptiker. 👍

File: scipy/spatial/distance.py (hunk 3 above)
Comment (ElDeveloper) [reply scipy/spatial/distance.py]: I am happy to change this for whatever you think might make more sense.

File: scipy/spatial/distance.py [hunk 4]
```diff
@@ -829,6 +829,18 @@ def jaccard(u, v, w=None):
     jaccard : double
         The Jaccard distance between vectors `u` and `v`.
 
+    Notes
+    -----
+    When both `u` and `v` lead to a `0/0` divsion i.e. there is no overlap
```
Comment (tylerjereddy) [inline scipy/spatial/distance.py]:
```suggestion
    When both `u` and `v` lead to a `0/0` division i.e. there is no overlap
```

Comment (tylerjereddy) [reply scipy/spatial/distance.py]: Caught one more typo here

Comment (ElDeveloper) [reply scipy/spatial/distance.py]: 👏 thanks @tylerjereddy

File: scipy/spatial/distance.py (hunk 3 above)
Comment (tylerjereddy) [reply scipy/spatial/distance.py]: Ternary-style operation is ok I suppose!

**Comment by tylerjereddy** [comment]
CI is all green and expert opinion has been weighed in so in this goes.

**Comment by tylerjereddy** [comment]
@ElDeveloper May be sensible to add a brief note of some sort about this change in the [release notes](https://github.com/scipy/scipy/wiki/Release-note-entries-for-SciPy-1.2.0), especially given that I didn't ask for a `versionchanged` type directive for the docs here.

**Comment by ElDeveloper** [comment]
Thanks @tylerjereddy. Just added a note to the Wiki. If you want I can also add a `versionchanged` directive in the docs.

**Comment by ilayn** [comment]
That would be great if you can add that. Often we need such quick information when resolving issues.

**Comment by ElDeveloper** [comment]
👍 just pushed https://github.com/scipy/scipy/pull/9446

**Pull request merged**